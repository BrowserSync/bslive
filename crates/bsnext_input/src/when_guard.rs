#[derive(Debug, Default, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum WhenGuard {
    #[default]
    Always,
    Never,
    Query {
        query: Vec<HasGuard>,
    },
}

#[test]
fn test_when_guard() {
    use insta::assert_debug_snapshot;
    #[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
    struct A {
        when: WhenGuard,
    }
    let input = r#"
    when:
      query:
        - not.has: "here"
    "#;
    let when: A = serde_yaml::from_str(input).expect("test");
    assert_debug_snapshot!(&when);
}

#[test]
fn test_when_guard_has() {
    use insta::assert_debug_snapshot;
    #[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
    struct A {
        when: WhenGuard,
    }
    let input = r#"
    when:
      query:
        - has: "here"
    "#;
    let when: A = serde_yaml::from_str(input).expect("test");
    assert_debug_snapshot!(&when);
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum HasGuard {
    /// A direct, exact match
    Is { is: String },
    /// Contains the substring
    Has { has: String },
    /// Contains the substring
    NotHas {
        #[serde(rename = "not.has")]
        not_has: String,
    },
}
