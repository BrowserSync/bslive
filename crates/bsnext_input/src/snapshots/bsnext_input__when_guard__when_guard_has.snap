---
source: crates/bsnext_input/src/when_guard.rs
expression: "&when"
---
A {
    when: Query {
        query: [
            Has {
                has: "here",
            },
        ],
    },
}
